### ISSUE TITLE
H265: Support HEVC in RTMP/SRT/HLS/WebRTC/DASH/GB28181

### ISSUE THREAD

**Issue by winlinvip** [ossrs/srs#465]
* [Overview](#overview)
* [Usage](#usage)
* [Status of HEVC in SRS](#status-of-hevc-in-srs)
* [FFmpeg Tools](#ffmpeg-tools)
* [MSE for HEVC](#mse-for-hevc)
* [Safari WebRTC](#safari-webrtc)
* [Thanks for Contributors](#thanks-for-contributors)
* [Known Issues](#known-issues)

<a name="overview"></a>
## [Overview](#overview)

The architecutre for SRS to support H.265(or HEVC):

```text
FFmpeg --RTMP(h.265)---> SRS ----RTMP/FLV/TS/HLS/WebRTC(h.265)--> Chrome/Safari
```

For live streaming:

* [Chrome 105+](https://caniuse.com/?search=HEVC) supports HEVC by default, see [this post](https://zhuanlan.zhihu.com/p/541082191).
  * You're able to play mp4 directly by H5 video, or by MSE if HTTP-FLV/HTTP-TS/HLS etc.
  * Please use [mpegts.js](https://github.com/xqq/mpegts.js) to play HTTP-TS with HEVC.
  * There is a plan for mpegts.js to support HTTP-FLV with HEVC, see [mpegts.js#64](https://github.com/xqq/mpegts.js/issues/64)
* FFmpeg or ffplay supports libx265
  * The `libx265` is available if install by `brew` on macOS.
  * Need some patch for HEVC over RTMP/FLV, see **[FFmpeg Tools](#ffmpeg-tools)** bellow.
* SRS also supports HEVC.
  * We're merging HEVC support to SRS 6.0
  * The original supports for HEVC is [srs-gb28181/feature/h265](https://github.com/ossrs/srs-gb28181/commits/feature/h265) by [runner365](https://github.com/runner365)

> Note: To check if your Chrome support HEVC, please open `chrome://gpu` and search `hevc`.

For WebRTC:

* Chrome does not support HEVC right now(2022.11), but supports AV1, please see [#2324](https://github.com/ossrs/srs/pull/2324)
* Safari supports HEVC if user enable it, please see this [section](#safari-webrtc)
* SRS also only supports AV1, because Chrome does not support HEVC yet.

<a name="usage"></a>
## [Usage](#usage)

Please make sure your SRS is `6.0.4+`, build with h265:

```bash
./configure --h265=on && make
```

Run SRS server:

```bash
env SRS_LISTEN=1935 SRS_DAEMON=off SRS_LOG_TANK=console SRS_HTTP_API_ENABLED=on \
  SRS_HTTP_SERVER_ENABLED=on SRS_VHOST_HTTP_REMUX_ENABLED=on \
  SRS_VHOST_HTTP_REMUX_MOUNT=[vhost]/[app]/[stream].flv \
  SRS_VHOST_HLS_ENABLED=on \
  ./objs/srs -e
```

> Note: Besides environment variables, you can also use `conf/hevc.flv.conf` or `conf/hevc.ts.conf` config files.
> Note: Recommend `conf/hevc.ts.conf` because TS is better for HEVC.

Build and patch FFmpeg, see [FFmpeg Tools](#ffmpeg-tools):

```bash
# For macOS
docker run --rm -it ossrs/srs:encoder ffmpeg -stream_loop -1 -re -i doc/source.flv \
  -acodec copy -vcodec libx265 -f flv rtmp://host.docker.internal/live/livestream

# For linux
docker run --net=host --rm -it ossrs/srs:encoder ffmpeg -stream_loop -1 -re -i doc/source.flv \
  -acodec copy -vcodec libx265 -f flv rtmp://127.0.0.1/live/livestream
```

> Note: Please change the ip `host.docker.internal` to your SRS's IP.

Play the HEVC live streams by:

* HTTP-FLV(by H5):  [http://localhost:8080/live/livestream.flv](http://localhost:8080/players/srs_player.html?autostart=true)
* HLS(by VLC or fflay): `http://localhost:8080/live/livestream.m3u8`

> Note: Please enable MPEG-DASH by `SRS_VHOST_DASH_ENABLED=on` then use VLC/ffplay to play stream `http://localhost:8080/live/livestream.mpd`

> Note: Please enable HTTP-TS by `SRS_VHOST_HTTP_REMUX_MOUNT=[vhost]/[app]/[stream].ts` then use H5/VLC/ffplay to play stream `http://localhost:8080/live/livestream.ts`

> Note: Please enable DVR MP4 by `SRS_VHOST_DVR_ENABLED=on SRS_VHOST_DVR_DVR_PATH=./objs/nginx/html/[app]/[stream].[timestamp].mp4` if want to covert live stream to MP4 file.

<img width="400" alt="image" src="https://user-images.githubusercontent.com/2777660/203508494-1bb95b6a-8cd8-4a2e-ba93-b02be8deab0e.png">

<img width="400" alt="image" src="https://user-images.githubusercontent.com/2777660/203508571-13b4d78f-6b5b-4aa0-8d0d-2636362ffe79.png">

<img width="400" alt="image" src="https://user-images.githubusercontent.com/2777660/203508601-3baf717e-b947-4544-bd0f-a26420603af4.png">

> Note: The detail about available protocols and tools for HEVC, please see [Status of HEVC in SRS](#status-of-hevc-in-srs).

> Note: The H5 player uses [mpegts.js](https://github.com/xqq/mpegts.js).

<a name="status-of-hevc-in-srs"></a>
## [Status of HEVC in SRS](#status-of-hevc-in-srs)

The status of protocols and HEVC:

* [x] PUSH HEVC over RTMP by FFmpeg, with none-official patch for FFmpeg. [v6.0.2](https://github.com/ossrs/srs/commit/178e40a5fc3cf0856ace914ae61696a73007f5bf)
* [x] PUSH HEVC over SRT by FFmpeg. [v6.0.20](https://github.com/ossrs/srs/pull/3366)
* [x] PUSH HEVC over RTMP by OBS. #3464 https://github.com/obsproject/obs-studio/pull/8522
* [x] PUSH HEVC over SRT by OBS. [v6.0.20](https://github.com/ossrs/srs/pull/3366)
* [x] PUSH HEVC over GB28181. [v6.0.25](https://github.com/ossrs/srs/pull/3408)
* [x] PULL HEVC over RTMP by FFmpeg, with none-official [patch](#ffmpeg-tools) for FFmpeg. [v6.0.2](https://github.com/ossrs/srs/commit/178e40a5fc3cf0856ace914ae61696a73007f5bf)
* [x] PULL HEVC over HTTP-FLV by FFmpeg, with none-official [patch](#ffmpeg-tools) for FFmpeg. [v6.0.2](https://github.com/ossrs/srs/commit/178e40a5fc3cf0856ace914ae61696a73007f5bf)
* [x] PULL HEVC over HTTP-TS by FFmpeg [v6.0.4](https://github.com/ossrs/srs/commit/70d5618979e5c8dc41b7cd87c78db7ca2b8a10e8)
* [x] PULL HEVC over HLS by FFmpeg [v6.0.11](https://github.com/ossrs/srs/commit/fff8d9863c3fba769b01782428257edf40f80a12)
* [x] PULL HEVC over MPEG-DASH  by FFmpeg [v6.0.14](https://github.com/ossrs/srs/commit/edba2c25f13c0fa915bd8e8093a4005df6077858)
* [x] PULL HEVC over SRT by FFmpeg. [v6.0.20](https://github.com/ossrs/srs/pull/3366)
* [x] PUSH HEVC over WebRTC by Safari. [v6.0.34](https://github.com/ossrs/srs/pull/3441)
* [x] PULL HEVC over WebRTC by Safari. [v6.0.34](https://github.com/ossrs/srs/pull/3441)
* [ ] PUSH HEVC over WebRTC by Chrome/Firefox
* [ ] PULL HEVC over WebRTC by Chrome/Firefox
* [x] Play HEVC over HTTP-TS by [mpegts.js](https://github.com/xqq/mpegts.js), by Chrome 105+ MSE, **NO WASM**. [v6.0.1](https://github.com/ossrs/srs/commit/7e02d972ea74faad9f4f96ae881d5ece0b89f33b)
* [x] Play pure video(no audio) HEVC over HTTP-TS by [mpegts.js](https://github.com/xqq/mpegts.js). [v6.0.9](https://github.com/ossrs/srs/commit/d5bf0ba2da30698e18700b210d2b12eed5b21d29)
* [x] Play HEVC over HTTP-FLV by [mpegts.js](https://github.com/xqq/mpegts.js), by Chrome 105+ MSE, **NO WASM**. [v6.0.1](https://github.com/ossrs/srs/commit/7e02d972ea74faad9f4f96ae881d5ece0b89f33b)
* [ ] Play HEVC over HLS by [hls.js](https://github.com/video-dev/hls.js)
* [ ] Play HEVC over MPEG-DASH by [dash.js](https://github.com/Dash-Industry-Forum/dash.js)
* [x] Play HEVC over HTTP-TS by ffplay, by offical release. [v6.0.4](https://github.com/ossrs/srs/commit/70d5618979e5c8dc41b7cd87c78db7ca2b8a10e8)
* [x] PULL HEVC over RTMP by ffplay, with none-official [patch](#ffmpeg-tools) for FFmpeg. [v6.0.2](https://github.com/ossrs/srs/commit/178e40a5fc3cf0856ace914ae61696a73007f5bf)
* [x] Play HEVC over HTTP-FLV by ffplay, with none-official [patch](#ffmpeg-tools) for FFmpeg. [v6.0.2](https://github.com/ossrs/srs/commit/178e40a5fc3cf0856ace914ae61696a73007f5bf)
* [x] Play pure video(no audio) HEVC by ffplay.
* [x] Play HEVC over HLS by ffplay. [v6.0.11](https://github.com/ossrs/srs/commit/fff8d9863c3fba769b01782428257edf40f80a12)
* [x] Play HEVC over MPEG-DASH by ffplay. [v6.0.14](https://github.com/ossrs/srs/commit/edba2c25f13c0fa915bd8e8093a4005df6077858)
* [x] Play HEVC over SRT by ffplay. [v6.0.20](https://github.com/ossrs/srs/pull/3366)
* [x] Play HEVC over HTTP-TS by VLC, by official release. [v6.0.4](https://github.com/ossrs/srs/commit/70d5618979e5c8dc41b7cd87c78db7ca2b8a10e8)
* [x] Play HEVC over SRT by VLC, by official. [v6.0.20](https://github.com/ossrs/srs/pull/3366)
* [x] Play pure video(no audio) HEVC by VLC.
* [ ] Play HEVC over RTMP by VLC.
* [ ] Play HEVC over HTTP-FLV by VLC.
* [x] Play HEVC over HLS by VLC. [v6.0.11](https://github.com/ossrs/srs/commit/fff8d9863c3fba769b01782428257edf40f80a12)
* [x] Play HEVC over MPEG-DASH by VLC. [v6.0.14](https://github.com/ossrs/srs/commit/edba2c25f13c0fa915bd8e8093a4005df6077858)
* [x] DVR HEVC to MP4/FLV file. [v6.0.14](https://github.com/ossrs/srs/commit/edba2c25f13c0fa915bd8e8093a4005df6077858)
* [x] HTTP API contains HEVC metadata.
* [ ] HTTP Callback takes HEVC metadata.
* [ ] Prometheus Exporter supports HEVC metadata.
* [ ] Improve coverage for HEVC.
* [x] Add regression/blackbox tests for HEVC.
* [ ] Supports benchmark for HEVC by [srs-bench](https://github.com/ossrs/srs-bench).
* [x] Support patched FFmpeg for SRS dockers: [CentOS7](https://github.com/ossrs/dev-docker/commit/0691d016adfe521f77350728d15cead8086d527d), [Ubuntu20](https://github.com/ossrs/dev-docker/commit/0e36323d15544ffe2901d10cfd255d9ef08fb250) and [Encoder](https://github.com/ossrs/dev-docker/commit/782bb31039653f562e0765a0c057d9f9babd1d1f).
* [x] Update [WordPress plugin SrsPlayer](https://github.com/ossrs/WordPress-Plugin-SrsPlayer) for HEVC.
* [ ] Update [srs-cloud](https://github.com/ossrs/srs-cloud) for HEVC.
* [ ] Edge server supports publish HEVC stream to origin.
* [ ] Edge server supprots play HEVC stream from origin.
* [ ] #3407

> Note: We're merging HEVC support to SRS 6.0, the original supports for HEVC is [srs-gb28181/feature/h265](https://github.com/ossrs/srs-gb28181/commits/feature/h265) by [runner365](https://github.com/runner365)

<a name="ffmpeg-tools"></a>
## [FFmpeg Tools](#ffmpeg-tools)

The FFmpeg in `ossrs/srs:encoder` or `ossrs/srs:6` is built with libx265 and patched with HEVC over RTMP support. So you're able to directly use:

```bash
docker run --rm -it ossrs/srs:encoder ffmpeg -re -i doc/source.flv -acodec copy \
  -vcodec libx265 -f flv rtmp://localhost/live/livestream
```

If you want to build from code, please read the bellow instructions.

The [specfication](https://github.com/ksvc/FFmpeg/wiki) and [usage](https://github.com/ksvc/FFmpeg/wiki/hevcpush) to support HEVC over RTMP or FLV. There is a [patch for FFmpeg 4.1/5.1/6.0](https://github.com/runner365/ffmpeg_rtmp_h265) from [runner365](https://github.com/runner365) for FFmpeg to support HEVC over RTMP or FLV. There is also a [patch](https://github.com/VCDP/CDN/blob/master/FFmpeg_patches/0001-Add-SVT-HEVC-FLV-support-on-FFmpeg.patch) from Intel for this feature.

Before build FFmpeg, we must build [libx264](https://www.videolan.org/developers/x264.html):

```bash
cd ~/git
git clone https://code.videolan.org/videolan/x264.git
cd ~/git/x264
./configure --prefix=$(pwd)/build --disable-asm --disable-cli --disable-shared --enable-static
make -j10
make install
```

And then [libx265](https://www.videolan.org/developers/x265.html):

```bash
cd ~/git
git clone https://bitbucket.org/multicoreware/x265_git.git
cd ~/git/x265_git/build/linux
cmake -DCMAKE_INSTALL_PREFIX=$(pwd)/build -DENABLE_SHARED=OFF ../../source
make -j10
make install
```

Now, please clone FFmepg and checkout to 5.1:

```bash
cd ~/git
git clone https://github.com/FFmpeg/FFmpeg.git
cd ~/git/FFmpeg && git checkout n5.1.2
```

Then, patch for [HEVC over RTMP/FLV](https://github.com/runner365/ffmpeg_rtmp_h265):

```bash
cd ~/git
git clone -b 5.1 https://github.com/runner365/ffmpeg_rtmp_h265.git
cp ~/git/ffmpeg_rtmp_h265/flv.h ~/git/FFmpeg/libavformat/
cp ~/git/ffmpeg_rtmp_h265/flv*.c ~/git/FFmpeg/libavformat/
```

Build ffmpeg with libx265:

```bash
cd ~/git/FFmpeg
env PKG_CONFIG_PATH=~/git/x264/build/lib/pkgconfig:~/git/x265_git/build/linux/build/lib/pkgconfig \
./configure \
  --prefix=$(pwd)/build \
  --enable-gpl --enable-nonfree --enable-pthreads --extra-libs=-lpthread \
  --disable-asm --disable-x86asm --disable-inline-asm \
  --enable-decoder=aac --enable-decoder=aac_fixed --enable-decoder=aac_latm --enable-encoder=aac \
  --enable-libx264 --enable-libx265 \
  --pkg-config-flags='--static'
make -j10
```

Push HEVC over RTMP to SRS:

```bash
./ffmpeg -stream_loop -1 -re -i ~/srs/doc/source.flv -acodec copy -vcodec libx265 \
  -f flv rtmp://localhost/live/livestream
```

Play HEVC over RTMP by ffplay:

```bash
./ffplay rtmp://localhost/live/livestream
```

It works.

<a name="mse-for-hevc"></a>
## [MSE for HEVC](#mse-for-hevc)

[MSE](https://caniuse.com/?search=mse) is a base technology for [mpegts.js](https://github.com/xqq/mpegts.js), [hls.js](https://github.com/video-dev/hls.js/) and [dash.js](https://github.com/Dash-Industry-Forum/dash.js).

Now [Chrome 105+](https://caniuse.com/?search=HEVC) supports HEVC by default, see [this post](https://zhuanlan.zhihu.com/p/541082191), which means, MSE(Chrome 105+) is available for HEVC.

You can verify this feature, by generating a HEVC mp4 file:

```bash
ffmpeg -i ~/git/srs/trunk/doc/source.flv -acodec copy \
  -vcodec libx265 -y source.hevc.mp4
```

> Note: Please make sure your FFmpeg is 5.0 and libx265 is enabled.

Open `source.hevc.mp4` in Chrome 105+ directly, it should works.

You can also move the file to SRS webserver:

```bash
mkdir -p ~/git/srs/trunk/objs/nginx/html/vod/
mv source.hevc.mp4 ~/git/srs/trunk/objs/nginx/html/vod
```

Then open by [srs-player](http://localhost:8080/players/srs_player.html?app=vod&stream=source.hevc.mp4&autostart=true)

<a name="safari-webrtc"></a>
## [Safari WebRTC](#safari-webrtc)

Safari supports WebRTC, if you enable it by:

* English version: `Develop > Experimental Features > WebRTC H265 codec`
* Chinese version: `开发 > 实验性功能 > WebRTC H265 codec`

Then open the url in safari, to publish or play WebRTC stream:

* Play [webrtc://localhost/live/livestream?codec=hevc](http://localhost:8080/players/rtc_player.html?codec=hevc)
* Publish [webrtc://localhost/live/livestream?codec=hevc](http://localhost:8080/players/rtc_publisher.html?codec=hevc)

Please follow other section to publish HEVC stream.

<a name="thanks-for-contributors"></a>
## [Thanks for Contributors](#thanks-for-contributors)

There is a list of commits and contributors about HEVC in SRS:

* [H265: For #1747, Support HEVC/H.265 in SRT/RTMP/HLS.](https://github.com/ossrs/srs-gb28181/commit/3ca11071b45495e82d2d6958e5d0f7eab05e71e5)
* [H265: For #1747, Fix build fail bug for H.265](https://github.com/ossrs/srs-gb28181/commit/e355f3c37228f3602c88fed68e8fe5e6ba1153ea)
* [H265: For #1747, GB28181 support h.265 (#2037)](https://github.com/ossrs/srs-gb28181/commit/b846217bc7f94034b33bdf918dc3a49fb17947e0)
* [H265: fix some important bugs (#2156)](https://github.com/ossrs/srs-gb28181/commit/26218965dd083d13173af6eb31fcdf9868b753c6)
* [H265: Deliver the right hevc nalu and dump the wrong nalu. (#2447)](https://github.com/ossrs/srs-gb28181/commit/a13b9b54938a14796abb9011e7a8ee779439a452)
* [H265: Fix multi nal hevc frame demux fail. #2494](https://github.com/ossrs/srs-gb28181/commit/6c5e6090d7c82eb37530e109c230cabaedf948e1)
* [H265: Fix build error #2657 #2664](https://github.com/ossrs/srs-gb28181/commit/eac99e19fba6063279b9e47272523014f5e3334a)
* [H265: Update mpegts demux in srt. #2678](https://github.com/ossrs/srs-gb28181/commit/391c1426fc484c990e4324a4ae2f0de900074578)
* [H265: Fix the stat issue for h265. (#1949)](https://github.com/ossrs/srs-gb28181/commit/b4486e3b51281b4c227b2cc4f58d2b06db599ce0)
* [H265: Add h265 codec written support for MP4 format. (#2697)](https://github.com/ossrs/srs-gb28181/commit/3175d7e26730a04b27724e55dc95ef86c1f2886e)
* [H265: Add h265 for SRT.](https://github.com/runner365/srs/commit/0fa86e4f23847e8a46e3d0e91e0acd2c27047e11)

We will merge some of these commits to SRS 6.0, but not all commits.

* [PULL HEVC over WebRTC by Safari. v6.0.34](https://github.com/ossrs/srs/pull/3441)
* [GB: Support H.265 for GB28181. v6.0.25 (#3408)](https://github.com/ossrs/srs/pull/3408)
* [H265: Support HEVC over SRT. v6.0.20 (#465) (#3366)](https://github.com/ossrs/srs/pull/3366)
* [H265: Support DVR HEVC stream to MP4. v6.0.14](https://github.com/ossrs/srs/pull/3360)
* HLS: Support HEVC over HLS. v6.0.11
* [HEVC: The codec information is incorrect. v6.0.5](https://github.com/ossrs/srs/issues/3271)
* FFmpeg support libx265 and HEVC over RTMP/FLV: [CentOS7](https://github.com/ossrs/dev-docker/commit/0691d016adfe521f77350728d15cead8086d527d), [Ubuntu20](https://github.com/ossrs/dev-docker/commit/0e36323d15544ffe2901d10cfd255d9ef08fb250) and [Encoder](https://github.com/ossrs/dev-docker/commit/782bb31039653f562e0765a0c057d9f9babd1d1f).
* [H265: Support HEVC over HTTP-TS. v6.0.4](https://github.com/ossrs/srs/commit/70d5618979e5c8dc41b7cd87c78db7ca2b8a10e8)
* [H265: Support parse multiple NALUs in a frame. v6.0.3](https://github.com/ossrs/srs/commit/f316e9a0de3a892d25f2d8e7efd28ee9334f5bd6)
* [H265: Support HEVC over RTMP or HTTP-FLV. v6.0.2](https://github.com/ossrs/srs/commit/178e40a5fc3cf0856ace914ae61696a73007f5bf)
* [H265: Update mpegts.js to play HEVC over HTTP-TS/FLV. v6.0.1](https://github.com/ossrs/srs/commit/7e02d972ea74faad9f4f96ae881d5ece0b89f33b)

<a name="known-issues"></a>
## [Known Issues](#known-issues)

1. HEVC over Safari WebRTC, only support WebRTC to WebRTC, doesn't support converting to RTMP.
2. Chrome/Firefox does not support HEVC, no any plan as I know.
3. Almost all browsers supports MSE, except iOS. HEVC over MSE requires hardware decoder.
4. Apart from mpegts.js, other H5 players such as hls.js/dash.js doesn't support HEVC.

**Comment by winlinvip** [comment]
赵文杰  20:38:44
Actually, for RTMP to support H.265, we need Adobe's approval.
赵文杰  20:38:58
There's no problem with HLS supporting H.265.
杨成立  20:39:02
As long as FFMPEG can play it, it's fine.
杨成立  20:39:06
Adobe is not an option.
赵文杰  20:39:32
Yes, SRS supports it easily, it depends on the RTMP parser.
杨成立  20:39:45
If VLC can play it, it's fine. You can provide a patch and see if VLC accepts it. If not, we can start a project.
赵文杰  20:39:51
Also, let's first create a streaming client.
杨成立  20:39:58
The delay is still too long for live streaming using HLS.
赵文杰  20:40:16
Okay, how about ffplay? Can it play it?
杨成立  20:40:24
Yes, it can.

`TRANS_BY_GPT3`

**Comment by djstava** [comment]
Doesn't srs 2.0 release support it?

`TRANS_BY_GPT3`

**Comment by victor9382** [comment]
Now I want to implement h265 HLS live streaming on version 2.0. I want to support h265 by modifying the code and use the latest ffmpeg to transcode the rtmp stream. However, it seems that h265 does not support flv, so I cannot directly transcode it like this (SRS transcodes to rtmp). In this case, is there a solution? The solution I thought of is to let SRS generate hls segments first, and then transcode each segment to h265.

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
The problem for SRS to support H.265+RTMP/FLV is no encoder support publish H.265 stream, so I filed a issue for [FFMPEG](http://trac.ffmpeg.org/ticket/6389). But it seems there is no reason for FFMPEG to support a private stuff that Adobe not update, so maybe I should ask Adobe about upgrading the RTMP/FLV protocol for H.265, which may take a very long time. Well, there seems no workaround for this issue.

It's very important for RTMP/FLV to support H.265, because for the interactive live streaming, which is 1-3s latency live streaming, the protocol plays an extremely important role and only RTMP/FLV is friendly to CDN. Neither HLS or DASH is possible to get 1-3s latency, even though I refine the HLS to HLS+(which is standard HLS with very dynamic small segments, about 0.8-1.2s segments), the latency of HLS is still 6s+. I want to tell everyone a rule **ALL segment based protocols such as HLS and DASH have large latency, they're impossible to get 1-3s latency**.

If your businesses require 1-3s latency and maybe migrate from H.264 to H.265, please comment to vote this issue [FFMPEG](http://trac.ffmpeg.org/ticket/6389). Remember **only when your business is requiring H.265+RTMP/FLV then vote it**, or FFMPEG developer will think all these votes are from sockpuppet accounts.

**Comment by huanghe2015** [comment]
@winlinvip, take a look at this link and you will understand why H.265 is not supported: [https://wenku.baidu.com/view/3905adeec850ad02df80414d.html](https://wenku.baidu.com/view/3905adeec850ad02df80414d.html)

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
@huanghe2015 FFMPEG does not support it not because of patent and licensing issues, but because Adobe has not updated the RTMP/FLV protocol. So if FFMPEG wants to support it, it would have to implement a non-standard protocol, which is not suitable for FFMPEG. In fact, other encoding methods, such as VP9 or the domestic standard mentioned in the article, would also have the same problem if FFMPEG were to support RTMP/FLV. So in reality, FFMPEG's lack of support for RTMP/FLV is like a death sentence, and we won't be able to work with it for real-time streaming.

For English user:
Nop, FFMPEG reject this feature not for fee, it's only because RTMP/FLV doesn't support H.265. So the right way is to wait for Adobe to upgrade the RTMP/FLV for H.265, or FFMPEG will be a private implmentation of H.265+RTMP/FLV. But this issue also exists when support other new codecs, such as VP9+RTMP/FLV, so it's a hard time for live streaming developer, especially for 1-3s interactive live streaming service.

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
There is a patch for FFMPEG to publish 265 over FLV: https://github.com/CDN-Union/Code/tree/master/flv265-ChinaNetCenter

FLV CodecID for H.265 by Kingsoft: https://github.com/CDN-Union/Code/tree/master/flv265-Kingsoft

**Comment by winlinvip** [comment]
Give up to support H.265 in FLV/RTMP, SRS3 will support it in HTTP-TS and HLS.

**Comment by fpzeng** [comment]
[FFmpeg patch](https://github.com/ksvc/FFmpeg/wiki) show both the document and code about the extensions for H.265/HEVC over flv, which defines the codec id 12 for HEVC  video packet for rtmp/flv live streaming.

**Comment by wellttllew** [comment]
FYI： 

- Aliyun CDN supports private RTMP codec id for HEVC

**Comment by winlinvip** [comment]
Regarding 265, my latest understanding is: 

1. TS is supported, which means HLS can package this codec. However, Safari, iOS, and Android browsers do not support this decoding. It only means that HLS packaging can support it.
2. RTMP is not supported because it is not updated by the official, which leads to FFMPEG only being able to support it through patching.
3. There are patent issues with 265, which may prevent it from being widely adopted. The current trend is to focus on codecs like AV1.

In summary, it is possible to support it, but it is estimated that it will only be supported within a small range.

`TRANS_BY_GPT3`

**Comment by yapingcat** [comment]
https://developer.apple.com/documentation/http_live_streaming/hls_authoring_specification_for_apple_devices

Apple's specifications require the use of fmp4 for h265. Although ts supports h265, currently, when transmitting h265 videos using the HLS protocol, fmp4 must be used.

`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
讨论请移步：#2133

**Comment by winlinvip** [comment]
Fixed by [feature/h265](https://github.com/ossrs/srs/tree/feature/h265)

**Comment by superemewill** [comment]
为啥我们的H265视频用srs的H265分支 播放不了，H264的视频可以播放

**Comment by superemewill** [comment]
> 为啥我们的H265视频用srs的H265分支 播放不了，H264的视频可以播放

是因为SRS支持H265，但是浏览器不支持

**Comment by winlinvip** [comment]
Plan to support HEVC in SRS 6.0

**Comment by bangjun** [comment]
用ffplay 播放h265时，出现卡顿，是什么原因呢

**Comment by bangjun** [comment]
When playing h265 with ffplay, there will be a freeze

**Comment by he852100** [comment]
I noticed that ByteDance company uses flv to transmit hevc, including tiktok and douyin.FLV has extremely low latency.

TS stream local buffering for up to 10 seconds, flv < 1s
